### ISSUE TITLE
possibly an odd question, but why use make and exclude windows test/contribs?

### ISSUE THREAD

**Issue by Pomax** [copy/v86#111]
This codebase uses javascript and clojure, so that's two universal scripting languages that work on all operating systems including windows, why not rewrite the make script so that it's a node-based bootstrapping script instead? Sure, you might need an `npm install` to grab some small dependencies for universal commandline utilities (although just [shx](https://www.npmjs.com/package/shx) might be more than enough), but it removes the cross-platform incompatibility that `make` and the use of bash shell instructions currently enforces.

There's already a package.json, so using npm scripts instead of a makefile would be pretty easy: just move each named step in the makefile into package.json, prepend some `shx` to most of the commands, and done: cross-platform compatibility without any headaches

**Comment by copy** [comment]
There's no particular reason why building on Windows isn't supported, besides nobody having asked for it yet ;-)

I'm not a fan of npm dependencies, but I don't think it would be bad to provide a separate mechanism for Windows users to build even if some functionality is duplicated. Contributions in that direction will be gladly accepted.

> This codebase uses javascript and clojure

Just a side note: It uses Closure Compiler, not clojure.

**Comment by jadr2ddude** [comment]
Also, the build should currently be possible with cygwin

**Comment by Pomax** [comment]
roger. In that case I guess this is a feature request from a windows user =)

(and cojure compiler: noted).

Cygwin might work, but is certainly not a very pleasant thing to have to install (and I say that as someone who's used it a lot over the last fifteen years), especially not now that windows is all about Powershell, which (of course, why not?) does not play super nice with bash syntax like ampersand and pipes (so you end up using `npm-run-all` which gives the nice `run-s` and `run-p` shorthand commands for running several npm scripts in sequence or in parallel without having to worry about the actual shell syntax for command sets)

**Comment by copy** [comment]
Just to clarify: You don't need to build the code to run it, and in fact I never do while developing. Just open `debug.html` and everything will load out of the box. The build step is only necessary for creating a minimized build for release.

**Comment by tracker1** [comment]
Closure Compiler has a node (transpiled) version available... so, npm could target .js files with a shebang instead of make files... which I and others in the JS sphere would probably prefer, but to each their own.

**Comment by Pomax** [comment]
but note that shebang is just as cross-platform-incompatible, windows does not respect shebangs in the slightest =)

**Comment by perlun** [comment]
Now with the Ubuntu for Windows (also known as Windows Subsystem for Linux), this should not be an issue. Just install WSL on Windows 10 and it should definitely be possible to build v86 on Windows only, it's just that you will actually do it in Ubuntu running in extremely thin virtualization on top of Windows. 😄 

More details from Microsoft [here](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide).

**Comment by Pomax** [comment]
Actually, the linux subsystem for windows is *not* useful there. As highlighted in https://msdn.microsoft.com/en-us/commandline/wsl/faq the WSL is so that you have access to common unix tools: you can compile tools so that they run in WSL, but compiling *in* WSL is almost guaranteed to fail.

**Comment by perlun** [comment]
@pomax, I think your assumption/interpretation of what they are writing there is incorrect. I just tried, it works fine:

```
plundberg@ECVAAWPLUN5-W10:~/v86$ make build/libv86.js
mkdir -p build
ls -lh build/libv86.js
ls: cannot access 'build/libv86.js': No such file or directory
Makefile:110: recipe for target 'build/libv86.js' failed
make: [build/libv86.js] Error 2 (ignored)
java -jar closure-compiler/compiler.jar \
        --js_output_file build/libv86.js\
        --define=DEBUG=false\
        --js lib/closure-base.js --generate_exports --externs src/externs.js --warning_level VERBOSE --jscomp_error accessControls --jscomp_error ambiguousFunctionDecl --jscomp_error checkEventfulObjectDisposal --jscomp_error checkRegExp --jscomp_error checkTypes --jscomp_error checkVars --jscomp_error conformanceViolations --jscomp_error const --jscomp_error constantProperty --jscomp_error deprecated --jscomp_error deprecatedAnnotations --jscomp_error duplicateMessage --jscomp_error es3 --jscomp_error es5Strict --jscomp_error externsValidation --jscomp_error fileoverviewTags --jscomp_error globalThis --jscomp_error inferredConstCheck --jscomp_error internetExplorerChecks --jscomp_error invalidCasts --jscomp_error misplacedTypeAnnotation --jscomp_error missingGetCssName --jscomp_error missingProperties --jscomp_error missingReturn --jscomp_error msgDescriptions --jscomp_error nonStandardJsDocs --jscomp_error suspiciousCode --jscomp_error strictModuleDepCheck --jscomp_error typeInvalidation --jscomp_error undefinedNames --jscomp_error undefinedVars --jscomp_error unknownDefines --jscomp_error unnecessaryCasts --jscomp_error visibility --use_types_for_optimization --summary_detail_level 3 --language_in ECMASCRIPT5_STRICT\
        --compilation_level SIMPLE\
        --language_in ECMASCRIPT6_STRICT --language_out ECMASCRIPT5_STRICT \
        --output_wrapper ';(function(){%output%})();'\
        --js src/const.js src/io.js src/main.js src/lib.js src/fpu.js src/ide.js src/pci.js src/floppy.js src/memory.js src/dma.js src/pit.js src/vga.js src/ps2.js src/pic.js src/rtc.js src/uart.js src/hpet.js src/acpi.js src/apic.js src/ioapic.js src/state.js src/ne2k.js src/virtio.js src/bus.js src/log.js src/config.js src/cpu.js src/translate.js src/modrm.js src/string.js src/arith.js src/misc_instr.js src/instructions.js src/debug.js src/elf.js\
        --js src/browser/screen.js src/browser/keyboard.js src/browser/mouse.js src/browser/serial.js src/browser/network.js src/browser/lib.js src/browser/starter.js src/browser/worker_bus.js src/browser/dummy_screen.js\
        --js lib/9p.js lib/filesystem.js lib/jor1k.js lib/marshall.js lib/utf8.js
0 error(s), 0 warning(s), 77.3% typed
ls -lh build/libv86.js
-rw-rw-rw- 1 plundberg plundberg 446K May 19 15:00 build/libv86.js
```

(Before you can build this, you must `apt-get update && apt-get install unzip default-jre make`)

Tested on Windows 10 Creators Update (10.0.15063). Yes, WSL is certainly not "perfect" but if you find issues, just report them at https://github.com/Microsoft/BashOnWindows. Microsoft _wants_ to make it better and they are improving it all the time.

**Comment by copy** [comment]
I think with the advent of WSL, Windows-specific builds are not necessary.
